### PR TITLE
Suggest to shorten temporary lifetime during method call inside generator

### LINKED ISSUE #67376: improve diagnostics for non-send Futures that might be solved by introducing temporary variables
In https://github.com/http-rs/tide/issues/370 we had a user report that they encountered a `Send` error similar to the ones discussed in https://github.com/rust-lang/rust/issues/64130. Thanks to the diagnostics introduced https://github.com/rust-lang/rust/pull/65345 on nightly we were able to correctly diagnose the root cause of the error.

But for many users this might still not be enough to figure out how to solve this, and we could improve by providing a hint.

## Current Diagnostics

2019-12-17-175740_1920x1080

```txt
   Compiling tide-demo v0.1.0 (/home/anon/Code/importcjj/tide-demo)
error: future cannot be sent between threads safely
  --> src/main.rs:47:27
   |
47 |         app.at("/submit").post(|req: tide::Request<State>| {
   |                           ^^^^ future is not `Send`
   |
   = help: the trait `std::marker::Sync` is not implemented for `(dyn futures_io::if_std::AsyncBufRead + std::marker::Send + 'static)`
note: future is not `Send` as this value is used across an await
  --> src/main.rs:49:29
   |
49 |                 let conn = &req.state().pool.get().await;
   |                             ---^^^^^^^^^^^^^^^^^^^^^^^^^- `req` is later dropped here
   |                             |
   |                             await occurs here, with `req` maybe used later
   |                             has type `&tide::request::Request<State>`
```

## Suggested Diagnostics

@estebank suggested we could add a suggestion here:

```txt
help: consider assigning `&req.state().pool` to a temporary variable
```

## Repro
- https://github.com/importcjj/tide-demo/blob/master/src/main.rs

## Refs
- https://github.com/rust-lang/rust/issues/64130
- https://github.com/http-rs/tide/issues/370

### PR DESCRIPTION
**Pull request by csmoe**
Closes https://github.com/rust-lang/rust/issues/67376
image

r? @estebank 
cc @tmandry @nikomatsakis

### CODE REVIEW

File: src/librustc/traits/error_reporting.rs [hunk 1]
```diff
@@ -2606,6 +2611,22 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+            if let Some(hir::Node::Expr(e)) = hir.find(parent) {
+                let method_span = hir.span(parent);
+                if tables.is_method_call(e) && is_ref {
+                    err.span_help(
+                        method_span,
+                        "consider moving this method call into a `let` \
+                        binding to create a shorter lived borrow",
+                    );
+                }
+            }
```
Comment (nikomatsakis) [inline src/librustc/traits/error_reporting.rs]: this looks like roughly what I expected to see =) seems good, from a quick glance!

File: src/librustc/ty/context.rs [hunk 2]
```diff
@@ -436,9 +436,9 @@ pub struct TypeckTables<'tcx> {
     /// entire variable.
     pub upvar_list: ty::UpvarListMap,
 
-    /// Stores the type, span and optional scope span of all types
+    /// Stores the type, expression, span and optional scope span of all types
     /// that are live across the yield of this generator (if a generator).
-    pub generator_interior_types: Vec<GeneratorInteriorTypeCause<'tcx>>,
+    pub generator_interior_types: Vec<(GeneratorInteriorTypeCause<'tcx>, Option<hir::HirId>)>,
```
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: er, wait, maybe we should move this `Option<hir::HirId>` into the `GeneratorInteriorTypeCause` structure?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by estebank** [comment]
After applying niko's suggestion, r=me.

**Comment by csmoe** [comment]
r=@estebank

**Pull request merged**